### DISCUSSION TITLE
Running behind a proxy

### DISCUSSION THREAD

**Discussion by tayden** [developmentseed/titiler#342] [Q&A] [answered]
I'd really like to get TiTiler running on an existing server, and have requests to the root path `/titiler` passed to the port where the TiTiler docker container is running. However, there's currently no way to specify the `root_path` parameter in the FastAPI instantiation call and so the app is required to run at the root location. (see FastAPI docs on [running behind a proxy](https://fastapi.tiangolo.com/advanced/behind-a-proxy/#setting-the-root_path-in-the-fastapi-app))

I'd suggest that titiler is enhanced by allowing users to set the `root_path` parameter given to the FastAPI instance via an environment variable so that it's dead simple to run the service behind a reverse proxy. It would be straightforward to default that root_path location to "/", so nothing anybody currently has running breaks.

**Comment by jaysnm** [comment]
Hi
I hope this gets an answer because I have fumbled on same issue for a couple of days without making any mojor progress. Below is what I have tried!

From [this issue](https://github.com/developmentseed/titiler/issues/95) I wrote my main.py script as follows:

```
from fastapi import FastAPI

from titiler.application.custom import templates
from titiler.application.routers import cog, mosaic, stac, tms
from titiler.application.version import __version__ as titiler_version

from starlette.requests import Request
from starlette.responses import HTMLResponse

from dotenv import dotenv_values

api_config = dotenv_values()

app = FastAPI(
    title=api_config.get('TILER_TITLE'),
    openapi_url="/api/openapi.json",
    description=api_config.get('TILER_DESC'),
    version=titiler_version,
)
app.include_router(
    cog.router, prefix=f"{api_config.get('TILER_SUB_PATH')}/cog", tags=["Cloud Optimized GeoTIFF"]
)
app.include_router(
    stac.router, prefix=f"{api_config.get('TILER_SUB_PATH')}/stac", tags=["SpatioTemporal Asset Catalog"]
)

app.include_router(
    mosaic.router, prefix=f"{api_config.get('TILER_SUB_PATH')}/mosaicjson", tags=["MosaicJSON"]
)
app.include_router(
    tms.router, prefix=f"{api_config.get('TILER_SUB_PATH')}/tileMatrixSets", tags=["TileMatrixSets"]
)
app.mount(api_config.get('TILER_SUB_PATH'), app)


@app.get(f"{api_config.get('TILER_SUB_PATH')}/healthz", description="Health Check", tags=["Health Check"])
def ping():
    """Health check."""
    return {"ping": "pong!"}


@app.get("/", response_class=HTMLResponse, include_in_schema=False)
def landing(request: Request):
    """TiTiler Landing page"""
    return templates.TemplateResponse(
        name="index.html", context={"request": request}, media_type="text/html",
    )
``` 

Where `TILER_SUB_PATH` denotes proxy sub path I wish to run titiler service. 
Simulation on `/docs` works fine for all endpoints embedding the service on another system fails with 
```
Access to fetch at 'http://domain.xx/titiler/cog/metadata/?url=http://domain.yy/stac/cogeo.tif&max_size=256' from origin 'http://domain.yy' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
``` 

Even after modifying cors middleware, it didn't work either!
```
from starlette.middleware.cors import CORSMiddleware
....

# Set all CORS enabled origins
app.add_middleware(
    CORSMiddleware,
    allow_origins=api_config.get('CORS_ORIGIN').split(","),
    allow_credentials=True,
    allow_methods=["GET"],
    allow_headers=["*"],
)
....
```

**Reply by tayden** [reply]
Hi @jaysnm, take a look at the root_path parameter you can pass to the FastAPI object https://fastapi.tiangolo.com/advanced/behind-a-proxy/#setting-the-root_path-in-the-fastapi-app

**Reply by jaysnm** [reply]
Hi @tayden 
I have followed the link and your PR #343 but still can't get it working. Appending `root_path` doesn't change access root url as I can load titiler on `/`. 
![image](https://user-images.githubusercontent.com/20385376/128666070-e126ca48-6c24-4200-9b53-97a379504e31.png)
While on the browser I have `/docs`, on the console it's printed that I have accessed `/titiler/docs` but swagger UI doesn't open because `openapi.json` cannot be found. 
![image](https://user-images.githubusercontent.com/20385376/128666334-c6fcba6d-ad8e-43a7-913a-af7160d2b814.png)
![image](https://user-images.githubusercontent.com/20385376/128666460-975188d2-16ef-4c27-86f8-e97e9036f30c.png)

The second output `/titiler/titiler` was generated when I tried `/titler` on browser. `/titiler/openapi.json` was generated when accessing `/openapi.json` on browser which outputs
![image](https://user-images.githubusercontent.com/20385376/128666766-bc9c7c58-ab25-4ece-8b60-bcf343477645.png)

**Reply by tayden** [reply]
If you're running on localhost, this would be the expected behaviour. The `root_path` parameter just tells helps FastAPI figure out the reverse proxy when you have the app running in production. The reverse proxy (e.g. Nginx or Apache) is responsible for adding the `/titiler` root_path (or whatever it is that you had set `root_path` to).

This section of the fastapi docs might be helpful to you https://fastapi.tiangolo.com/advanced/behind-a-proxy/#about-root_path

**Reply by lsntdev** [reply]
Hello Guys, I think I have the same issue when trying to use an existing nginx config, I tried to integrate it with geonode. Is there any way to do this without the change in the source code like here https://github.com/developmentseed/titiler/issues/95. I asked an help  [here](https://stackoverflow.com/questions/76615839/how-to-setup-titiler-for-customized-root-path-to-be-used-on-an-existing-nginx-co).

**Comment by jaysnm** [comment, marked as answer]
Went through the links once more only to realize it was a shortcoming on my end. Thanks a lot.
